app/views/tracks/show/joined/_uncompleted_exercises_section.html.haml

```diff
@@ -0,0 +1,12 @@
+- return if user_track.uncompleted_exercises.blank?
```

Comment:
You'll clearly have thought of this, but what was the reason to make this the responsibility of this view instead of the caller?

Reply:
Excellent question.

So the main answer is that then we don't have to worry about having that check in lots of files whenever the partial is called. We can encapsulate that logic in one place. 

As a general rule it's also good because:
```
= render "some_partial", collection: some_array
```
is 3x faster than
```
- some_array.each do |item|
  = render "some_partial", item: item
```
despite the fact they have the same output.

That means if we wanted to add a check outside the partial, we'd have to use the second method of rendering. Whereas if we do any checks inside the partial, we can use the first (faster) method.

---

All that said I actually dislike the partial rendering thing quite a lot. Experience has taught (hurt) that it's easy to make mistakes with. So I prefer the `ViewComponent` approach we've invented/taken for v3, where we can encapsulate the logic and everything in one place, and with it faster than haml parsing.

But for situations like this, where we're effectively just breaking up one big template into it's sections, the `ViewComponent` approach would be overkill.

Reply:
The other approach would be to pre-filter `some_array`, and in this case that's not necessarily more performant or even better at all (just to give you the other side of the coin @ErikSchierboom).

Partial rendering (or rather template rendering) just has performance caveats that are "easy" to run in to; to elaborate what can hurt. In other words: partials aren't bad but are easy to do in a non-performant way, which makes them meh.